src/NLSModels.jl

```diff
@@ -0,0 +1,301 @@
+export AbstractNLSModel, nls_meta, NLSCounters, nls_counters, reset!,
+       residual, residual!, jac_residual, jprod_residual, jprod_residual!,
+       jtprod_residual, jtprod_residual!, jac_op_residual, jac_op_residual!,
+       hess_residual, hprod_residual, hprod_residual!, hess_op_residual,
+       hess_op_residual!, NotImplementedError
+
+abstract AbstractNLSModel <: AbstractNLPModel
+
+type NLSCounters
+  counters :: Counters
+  neval_residual        :: Int
+  neval_jac_residual    :: Int
+  neval_jprod_residual  :: Int
+  neval_jtprod_residual :: Int
+  neval_hess_residual   :: Int
+  neval_hprod_residual  :: Int
+
+  function NLSCounters()
+    return new(Counters(), 0, 0, 0, 0, 0, 0)
+  end
+end
+
+nls_counters(nls :: AbstractNLSModel) = nls.counters
+
+function sum_counters(c :: NLSCounters)
+  s = sum_counters(c.counters)
+  for field in fieldnames(NLSCounters)
+    if field == :counters; continue; end
```

Comment:
Ah yes. I was using `if A; B end` instead of `A && B` for some time. But now that I look back, I can't see a good reason why.